### PR TITLE
Task 11

### PR DESCRIPTION
**Pull request by Xodoc**

### CODE REVIEW

File: ConsoleApp/BL/Services/CityService.cs [hunk 1]
```diff
@@ -35,20 +35,30 @@ public async Task<List<CityDTO>> CheckAndCreateCitiesAsync()
 
             var newCities = new List<City>();
 
-            foreach (var city in oldCities) 
+            foreach (var city in oldCities)
             {
                 if (_config.CityNames.Contains(city.CityName) == false)
```
Comment (denja2) [inline ConsoleApp/BL/Services/CityService.cs:40]: P2 iterate by cities from config

File: ConsoleApp/IntegrationTests/DI.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+using Shared.Config;
+using Shared.Extensions;
+
+namespace IntegrationTests
+{
+    public static class DI
+    {
+        public static IServiceCollection AddDI(this IServiceCollection services)
+        {
+            services.AddDbContext<ApplicationDbContext>(options => options.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=mentorshipdb;Trusted_Connection=True;"));
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/DI.cs:19]: To config

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        [Theory]
+        [InlineData("Minsk", "12.03.2022", "16.03.2022")]
+        public async void GetWeatherHistoryAsync_WhenSendingCorrectData_GettingWeatherHistory(string cityName, string from, string to)
+        {
+            //Arrange
+            var expected = new WeatherHistoryDTO { CityId = 1, Timestamp = DateTime.Parse(from), Temp = -2.14 };
+            var data = _mapper.Map<WeatherHistory>(expected);
+
+            //Act
+            await _context.WeatherHistories.AddAsync(data);
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:52]: Arrange

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+            GC.SuppressFinalize(this);
+        }
+
+        [Theory]
+        [InlineData("Minsk", "12.03.2022", "16.03.2022")]
+        public async void GetWeatherHistoryAsync_WhenSendingCorrectData_GettingWeatherHistory(string cityName, string from, string to)
+        {
+            //Arrange
+            var expected = new WeatherHistoryDTO { CityId = 1, Timestamp = DateTime.Parse(from), Temp = -2.14 };
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:48]: add more cities

File: ConsoleApp/Tests/Services/WeatherHistoryServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            _mapper = new MapperConfiguration(cfg => cfg.AddProfile(new WeatherHistoryProfile())).CreateMapper();
+            _fixture = (Fixture)new Fixture().Customize(new AutoMoqCustomization());
+            _weatherHistoryRepositoryMock = _fixture.Freeze<Mock<IWeatherHistoryRepository>>();
+            _weatherRepository = _fixture.Freeze<IWeatherRepository>();
+            _weatherHistoryService = new WeatherHistoryService(_weatherHistoryRepositoryMock.Object, _weatherRepository, _mapper);
+        }
+
+        [Theory]
+        [InlineData("Minsk", "16.03.2022", "17.03.2022")]
```
Comment (denja2) [inline ConsoleApp/Tests/Services/WeatherHistoryServiceTest.cs:37]: Test add method

File: ConsoleApp/WebAPI/Controllers/WeatherHistoryController.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+        public WeatherHistoryController(IWeatherHistoryService weatherHistoryService)
+        {
+            _weatherHistoryService = weatherHistoryService;
+        }
+
+        [HttpGet("getWeatherHistory")]
+        public async Task<IActionResult> GetWeatherHistory([FromQuery] WeatherHistoryRequest request)
+        {
+            return Ok(await _weatherHistoryService.GetWeatherHistoriesAsync(request.CityName, DateTime.Parse(request.From), DateTime.Parse(request.To)));
```
Comment (denja2) [inline ConsoleApp/WebAPI/Controllers/WeatherHistoryController.cs:21]: Use datetime arguments

**Review by denja2 — CHANGES_REQUESTED** [review]

File: ConsoleApp/BL/Services/CityService.cs (hunk 1 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/CityService.cs:40]: done

File: ConsoleApp/IntegrationTests/DI.cs (hunk 2 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/DI.cs:19]: done

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs (hunk 4 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:48]: done

File: ConsoleApp/WebAPI/Controllers/WeatherHistoryController.cs (hunk 6 above)
Comment (Xodoc) [reply ConsoleApp/WebAPI/Controllers/WeatherHistoryController.cs:21]: done

File: ConsoleApp/Tests/Services/WeatherHistoryServiceTest.cs (hunk 5 above)
Comment (Xodoc) [reply ConsoleApp/Tests/Services/WeatherHistoryServiceTest.cs:37]: done

File: ConsoleApp/BL/Services/CityService.cs [hunk 7]
```diff
@@ -35,26 +35,17 @@ public async Task<List<CityDTO>> CheckAndCreateCitiesAsync()
 
             var newCities = new List<City>();
 
-            foreach (var city in oldCities)
+            foreach (var name in _config.CityNames)
             {
-                if (_config.CityNames.Contains(city.CityName) == false)
+                if (oldCities.Any(x => x.CityName == name))
```
Comment (denja2) [inline ConsoleApp/BL/Services/CityService.cs:40]: P2 first or default

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs [hunk 8]
```diff
@@ -45,17 +46,23 @@ public async void Dispose()
         public async void GetWeatherHistoryAsync_WhenSendingCorrectData_GettingWeatherHistory(string cityName, string from, string to)
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:46]: P3 remove inline data

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs [hunk 9]
```diff
@@ -75,5 +82,22 @@ public async void GetWeatherHistoryAsync_WhenSendingIncorrectData_GettingWeather
             //Asssert
             Assert.Equal(expectedList, actualResult, new WeatherHistoryDTOComparator());
         }
+
+        [Theory]
```
Comment (denja2) [inline ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:86]: CheckAndCreateCitiesAsync() in unit test

**Review by denja2 — CHANGES_REQUESTED** [review]

File: ConsoleApp/BL/Services/CityService.cs (hunk 7 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/CityService.cs:40]: done

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs (hunk 8 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs:46]: done

File: ConsoleApp/WebAPI/Controllers/WeatherController.cs [hunk 10]
```diff
@@ -15,6 +15,13 @@ public WeatherController(IWeatherService weatherService)
             _weatherService = weatherService;
         }
 
+        /// <summary>
+        /// Getting current weather by city name
+        /// </summary>
+        /// <param name="weather"></param>
+        /// <returns></returns>
+        /// <response code="200">Success</response>
+        /// <response code="400">If send data is incorrect</response>
```
Comment (denja2) [inline ConsoleApp/WebAPI/Controllers/WeatherController.cs:24]: Whats about 500?

**Review by denja2 — CHANGES_REQUESTED** [review]

File: ConsoleApp/WebAPI/Controllers/WeatherController.cs (hunk 10 above)
Comment (Xodoc) [reply ConsoleApp/WebAPI/Controllers/WeatherController.cs:24]: done

File: ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs (hunk 3 above)
Comment (Xodoc) [reply ConsoleApp/IntegrationTests/Services/WeatherHistoryServiceIntegrationTest.cs]: done

**Pull request merged**